clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -109,6 +109,23 @@ static bool SYCLCUDAIsSYCLDevice(const clang::LangOptions &LangOpts) {
   return LangOpts.SYCLIsDevice && LangOpts.CUDA && !LangOpts.CUDAIsDevice;
 }
 
+static bool isSyclType(QualType Ty, SYCLTypeAttr::SYCLType TypeName) {
```

Comment:
I think instead of defining and using this function you can just do
```
    if (const auto *Attr = ED->getAttr<SYCLTypeAttr>())
      if (Attr->getType() == SYCLTypeAttr::SYCLType::host_pipe)
         // do smth

```
Just because at codegen level it is expected that all templates are resolved and semantic checks are performed (so no need to double check template decl and whether underlying type is a record).

Reply:
Example https://github.com/intel/llvm/blob/4afeea5c9f7a2f564c39c25cbe913c1cd8c5082b/clang/lib/CodeGen/CodeGenTypes.cpp#L337